firedrake/mg/kernels.py

```diff
@@ -310,7 +311,8 @@ def restrict_kernel(Vf, Vc):
         return cache[key]
     except KeyError:
         mesh = extract_unique_domain(coordinates)
-        evaluate_code = compile_element(firedrake.TestFunction(Vc), Vf)
+        v = firedrake.TestFunction(Vc.dual()) if isinstance(Vc, FiredrakeDualSpace) else firedrake.TestFunction(Vc)
+        evaluate_code = compile_element(v, Vf)
```

Comment:
I am not entirely sure this is correct.

Reply:
At this point we are guaranteed that `Vc` is a `DualSpace`.

Reply:
Oh, this is not the case for more than 1 refinement per level.